### PR TITLE
#159987708 Like and dislike functionality

### PR DESCRIPTION
**Pull request by iamuchejude**
### What does this PR do?
Add like and dislike feature for articles

### Description of the task?
* Create endpoints for like and dislike functionality
* Create endpoint to get rating/reactions for an article
* Write relevant test for like and dislike feature

### How to manually test
* Clone the repository
* cd to `thor-ah`
* Setup postgres database
* Create `.env` using the format in `.env.sample`
* Run `npm install`
* Run `npm run start:dev`
* Test the following endpoints using postman or any other available alternative software
**POST**  /api/articles/:article_slug/reactions
**GET** /api/articles/:article_slug/reactions 

### What are the relevant Pivotal Tracker Story?
[#159987708](https://www.pivotaltracker.com/story/show/159987708)

### CODE REVIEW

File: server/controllers/likeDislike.js [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+              res.status(200).json({
+                status: 'success',
+                reaction: {
+                  likes: 0,
+                  dislikes: 0,
+                }
+              });
+            }
+            likesDislikes.forEach((likeAndDislike) => {
+              if (likeAndDislike.status === 'likedßß') {
```
Comment (emekafredy) [inline server/controllers/likeDislike.js]: On line 152, shouldn't it be 'liked' instead of 'likedBB'?

File: server/models/likedislike.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+module.exports = (sequelize, DataTypes) => {
```
Comment (emekafredy) [inline server/models/likedislike.js]: You have two identical models for 'likedislike'. Same applies to migrations

**Comment by waleCloud** [comment]
@iamuchejude 

looking through your commit messages, i can see upto 4 commits that has `New Update` as the commit message attached to it.

I will suggest you be more descriptive about your messages, take a look at @d-beloved 's commit from the PR he raised.
I couldn't help but to point this out while looking through the codebase.

**Review by CEOehis — APPROVED** [review]
LGTM

**Review by emekafredy — APPROVED** [review]

**Review by emekafredy — CHANGES_REQUESTED** [review]
Sorry, I think you have conflicts to resolve in 'server/controllers/users.js'

File: server/controllers/likeDislike.js [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        }
+        const articleId = article.id;
+
+        LikesDislikes.findOne({
+          where: {
+            userId,
+          }
+        })
+          .then((likeOrDislike) => {
+            if (likeOrDislike === null || likeOrDislike === undefined) {
```
Comment (aknwosu) [inline server/controllers/likeDislike.js]: if (!likeOrDislike) {

File: server/controllers/likeDislike.js [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+      return res.status(400).json({
+        status: 'error',
+        message: 'Please provide a valid article slug.',
+      });
+    }
+
+    if (!reaction || reaction.trim().length < 1) {
+      return res.status(400).json({
+        status: 'error',
+        message: 'Please provide a valid reaction .',
```
Comment (aknwosu) [inline server/controllers/likeDislike.js]: message: 'No reaction provided.',

**Review by emekafredy — APPROVED** [review]
LGTM

File: server/migrations/20180907145330-create-likes-dislikes.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    userId: {
+      type: Sequelize.INTEGER,
+      allowNull: false,
+    },
+    articleId: {
+      type: Sequelize.INTEGER,
+      allowNull: false,
+    },
+    status: {
+      type: Sequelize.ENUM('liked', 'disliked'),
```
Comment (d-beloved) [inline server/migrations/20180907145330-create-likes-dislikes.js]: you can change the ENUM value to `1` and `0`, so when an article is liked, it takes 1 and when it is disliked, it takes `0`

**Review by lauragift21 — CHANGES_REQUESTED** [review]
For naming your models, you can add underscore between the words making it likes_dislikes.js.

**Review by d-beloved — CHANGES_REQUESTED** [review]

File: server/migrations/20180907145330-create-likes-dislikes.js (hunk 5 above)
Comment (iamuchejude) [reply server/migrations/20180907145330-create-likes-dislikes.js]: Thanks for that. I have implemented the feedback

**Review by d-beloved — APPROVED** [review]

**Comment by lauragift21** [comment]
LGTM

**Review by lauragift21 — APPROVED** [review]

File: server/controllers/likeDislike.js [hunk 6]
```diff
@@ -0,0 +1,167 @@
+import db from '../models';
+
+const { Article, LikesDislikes } = db;
+
+/**
+ * @class likeDislike
+ */
+class likesDislikesController {
+  /**
+   * @desc Like an article
```
Comment (aknwosu) [inline server/controllers/likeDislike.js]: misleading comment

File: server/controllers/likeDislike.js [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+   * @param {obj} req request body
+   * @param {obj} res response body
+   * @param {obj} next next action
+   * @return {obj} returns an object
+   */
+  static likeOrDislike(req, res, next) {
+    const { articleSlug } = req.params;
```
Comment (aknwosu) [inline server/controllers/likeDislike.js]: use article id as the check

File: server/controllers/likeDislike.js [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    if ((reaction === 'like') !== true) {
+      if ((reaction === 'dislike') !== true) {
+        return res.status(400).json({
+          status: 'error',
+          message: "Reaction must either be 'like' or 'dislike'",
+        });
+      }
+    }
```
Comment (aknwosu) [inline server/controllers/likeDislike.js]: duplicate check. just keep the second one. i.e. 
if ((reaction === 'like') !== true) {
      if ((reaction === 'dislike') !== true) {
        return res.status(400).json({
          status: 'error',
          message: "Reaction must either be 'like' or 'dislike'",
        });
      }
    }

File: server/controllers/likeDislike.js [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+
+    const status = reaction === 'like' ? '1' : '0';
+
+    Article.findOne({
+      where: {
+        slug: articleSlug,
```
Comment (aknwosu) [inline server/controllers/likeDislike.js]: use articleID

File: server/controllers/likeDislike.js [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+          .then((likeOrDislike) => {
+            if (!likeOrDislike) {
+              LikesDislikes.create({
+                articleId,
+                userId,
+                reaction: status,
+              })
+                .then(() => res.status(200).json({
+                  status: 'success',
+                  message: `Article ${reaction}d successfully.`,
```
Comment (aknwosu) [inline server/controllers/likeDislike.js:78]: nice

File: server/controllers/likeDislike.js [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+   * @return {obj} returns an object
+   */
+  static getLikesDislikes(req, res, next) {
+    const { articleSlug } = req.params;
+    const reactions = { likes: 0, dislikes: 0 };
+    if (!articleSlug || articleSlug.trim().length < 1) {
+      return res.status(400).json({
+        status: 'error',
+        message: 'Please provide a valid article slug.',
+      });
```
Comment (aknwosu) [inline server/controllers/likeDislike.js]: No valid article ID provided

File: server/controllers/likeDislike.js [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+            message: 'Article was not found.',
+          });
+        }
+        LikesDislikes.findAll({
+          where: {
+            articleId: article.id,
+          },
+        })
+          .then((likesDislikes) => {
+            if (likesDislikes === null) {
```
Comment (aknwosu) [inline server/controllers/likeDislike.js]: if (!likesDislikes) {

File: server/controllers/likeDislike.js [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+      }
+    })
+      .then((article) => {
+        if (!article) {
+          return res.status(404).json({
+            status: 'error',
+            message: 'Article was not found.',
+          });
+        }
+        LikesDislikes.findAll({
```
Comment (aknwosu) [inline server/controllers/likeDislike.js]: let likes,
let dislikes
LikesDislikes.findAll({
where: {
            articleId: article.id,
            reaction: 0}).then((allLikes) => {if (!allLikes} { likes === 0} else {likes = allLikes.Length})


same for dislikes. return data: {likes}, {dislikes}
          },

**Review by aknwosu — CHANGES_REQUESTED** [review]

File: server/controllers/likeDislike.js [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+            res.status(200).json({
+              status: 'success',
+              reactions,
+            });
+          });
+      }).catch(err => next(err));
+  }
+}
+
+export default likesDislikesController;
```
Comment (aknwosu) [inline server/controllers/likeDislike.js:225]: check if a user liked or disliked an article so it can be used on the front end

**Review by emekafredy — CHANGES_REQUESTED** [review]
On line 24 of your likes-dislikes controller, message should be a property of an error object, as we have agreed. something like this:
error: {
   message: 'error message here'
}

Same applies for lines 31, 39, 54, 117, 130, 173 and 185.

**Review by emekafredy — APPROVED** [review]

**Review by emekafredy — COMMENTED** [review]
LGTM

File: server/controllers/likeDislike.js [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+                dislikes: results[1].length,
+              },
+            });
+          }).catch(err => next(err));
+      }).catch(err => next(err));
+  }
+
+  /**
+   * @desc Get all reaction status of an article for a particular user
+   * @param {obj} req request body
```
Comment (aknwosu) [inline server/controllers/likeDislike.js]: misleading comment

File: server/controllers/likeDislike.js [hunk 16]
```diff
@@ -0,0 +1,227 @@
+import db from '../models';
+
```
Comment (aknwosu) [inline server/controllers/likeDislike.js]: import { Article, LikesDislikes } from '../models';

File: server/controllers/likeDislike.js [hunk 17]
```diff
@@ -0,0 +1,227 @@
+import db from '../models';
```
Comment (aknwosu) [inline server/controllers/likeDislike.js]: import { Article, LikesDislikes } from '../models';

**Review by aknwosu — CHANGES_REQUESTED** [review]

File: server/controllers/likeDislike.js (hunk 15 above)
Comment (iamuchejude) [reply server/controllers/likeDislike.js]: Okay...I will update it

File: server/controllers/likeDislike.js (hunk 17 above)
Comment (iamuchejude) [reply server/controllers/likeDislike.js]: Thanks. I will do the needful

**Pull request merged**